### ISSUE TITLE
Emscipten/WebGL support and the lack of glMapBufferRange/glUnmapBuffer

### ISSUE THREAD

**Issue by eugeneko** [DiligentGraphics/DiligentCore#404]
Pretty much the subject.

WebGL 2.0 specs explicitly lack glMapBufferRange and glUnmapBuffer: https://registry.khronos.org/webgl/specs/latest/2.0/#5.14

Maybe there can be a workaround on Diligent level?
These are only these two functions missing, and they are 100% workaroundable via non-mapping API.

**Comment by TheMostDiligent** [comment]
I don't understand what the question/suggestion really is.
Diligent supports buffer mapping/unmapping through `IDeviceContxt::MapBuffer`/`IDeviceContxt::UnmapBuffer` in Emscripten.

**Comment by eugeneko** [comment]
>Diligent supports buffer mapping/unmapping through IDeviceContxt::MapBuffer/IDeviceContxt::UnmapBuffer in Emscripten

Diligent uses functions `glMapBufferRange` and `glUnmapBuffer` in `BufferGLImpl::MapRange/Unmap` which is called from `IDeviceContext::MapBuffer/UnmapBuffer`. These functions _don't exist_ in WebGL.

Therefore, Diligent can only work if I enable GLES3.0 emulation in Emscripten, instead of using WebGL2.
If I try to use Emscripten in WebGL2 subset, I will get "unresolved symbol" errors for `glMapBufferRange` and `glUnmapBuffer`

**Comment by eugeneko** [comment]
tl;dr: I wonder if it's possible to use Diligent with "WebGL-friendly subset of OpenGL ES 3.0" instead of "OpenGL ES 3.0 emulation", going by this article: https://emscripten.org/docs/porting/multimedia_and_graphics/OpenGL-support.html

If it's hard to do, then it's probably not worth it. But if it's only these two functions...

**Comment by TheMostDiligent** [comment]
What is the original problem that you are trying to solve?

**Comment by TheMostDiligent** [comment]
> These functions (glMapBufferRange and glUnmapBuffer) don't exist in WebGL.

Sorry, I missed this comment. I think it should be possible to implement buffer mapping through `glBufferData`. I am wondering though, how many browsers don't support GLES3.0 emulation.

**Comment by eugeneko** [comment]
> I am wondering though, how many browsers don't support GLES3.0 emulation.

AFAIK emulation is done by Emscripten, not by browsers.
My only real concern here is that there is no option to "emulate buffer mapping and nothing else".
Emscripten has two modes: high-performance "pure" WebGL2, and emulated GLES3 with possible performance penalties.
And I don't know the scope of this "emulation" and how much it does affect performance in practice.
Is it "don't pay for what you don't use" or does it incur additional penalties?..

**Comment by TheMostDiligent** [comment]
> Is it "don't pay for what you don't use" or does it incur additional penalties?..

I don't know. Since GLES3 is largely similar to WebGL2, I don't see any reason why that would be substantially slow.
I also don't see the point in re-implementing something that has already been done unless that implementation has been shown to be a real problem.

**Comment by eugeneko** [comment]
I found curious issue which may be related to this topic.
I am getting this warning in RenderDoc when using OpenGL:

> Buffer performance warning: Buffer object CameraCB#4116606499 (bound to GL_UNIFORM_BUFFER (0), and GL_UNIFORM_BUFFER_EXT, usage hint is GL_DYNAMIC_DRAW) is being copied/moved from VIDEO memory to SYSTEM HEAP memory

I am doing `IDeviceContext::MapBuffer(MAP_WRITE, MAP_FLAG_DISCARD)` for `USAGE_DYNAMIC` buffer on my side, so it is as "optimal" as it can be.

On the other hand, Apple suggest to use glMapBufferRange on their devices:

https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/OpenGLES_ProgrammingGuide/TechniquesforWorkingwithVertexData/TechniquesforWorkingwithVertexData.html

> In iOS, GL_DYNAMIC_DRAW and GL_STREAM_DRAW are equivalent. You can use the glBufferSubData function to update buffer contents, but doing so incurs a performance penalty because it flushes the command buffer and waits for all commands to complete. Double or triple buffering can reduce this performance cost somewhat. (...) For better performance, use the glMapBufferRange function in OpenGL ES 3.0 or the corresponding function provided by the [EXT_map_buffer_range](http://www.khronos.org/registry/gles/extensions/EXT/EXT_map_buffer_range.txt) extension in OpenGL ES 2.0 or 1.1.

**Comment by TheMostDiligent** [comment]
OpenGL backend should really only be used on Web. On other platforms any other backend will perform much better. On Apple, OpenGL has been deprecated for over a decade now. Metal (or Vulkan + MoltenVK) should be used on Apple platforms instead.

**Comment by TheMostDiligent** [comment]
Buffer mapping in WebGL was reworked in 55c46aa1b97ea3e0067c3dc65b024c18925f56a4
